### ISSUE TITLE
How do you listen for dirty/pristine state of the form control?

### ISSUE THREAD

**Issue by slavafomin** [angular/angular#17736]
I've posted this question on StackOverflow recently: [How do I know when custom form control is marked as pristine in Angular?](https://stackoverflow.com/questions/44730711/how-do-i-know-when-custom-form-control-is-marked-as-pristine-in-angular).

After further investigation, I've managed to get access to both `NgModel` and `FormControl` from my custom form component, but I can't find any observable that I can use to actually listen for dirty/pristine state of the form control. The `ControlValueAccessor` is not providing any hooks for this task either.

Is it possible to watch for dirty/pristine state of the form control somehow? There must be an observable for this?

**Comment by slavafomin** [comment]
After looking at the source code, it looks like [`markAsPristine()`](https://github.com/angular/angular/blob/master/packages/forms/src/model.ts#L288) method is very simple and doesn't emit anything. Is there alternative ways I can use to listen for control's dirty/pristine state?

Also, are you willing to mark this as a feature request? I think there should be a way to listen for those changes in component code, without using bindings in templates.

**Comment by Toxicable** [comment]
Hello, we reviewed this issue and determined that it doesn't fall into the bug report or feature request category. This issue tracker is not suitable for support requests, please repost your issue on [StackOverflow](http://stackoverflow.com/) using tag `angular`.

If you are wondering why we don't resolve support issues via the issue tracker, please [check out this explanation](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-got-a-question-or-problem).

**Comment by Toxicable** [comment]
You can listen to an `AbstractControl#statusChanges` eg, `control.statusChanges.map( state => if(state == 'VLID" { //do something })`

**Comment by slavafomin** [comment]
> You can listen to an AbstractControl#statusChanges eg, control.statusChanges.map( state => if(state == 'VLID" { //do something })

You will need to use `subscribe()` instead of `map()`, otherwise cold Observable wouldn't emit anything. At least in my experience.

Also, the documentation for `statusChanges` observable states: `Emits an event every time the validation status of the control is re-calculated.`.

So how it would help me with watching for **dirty/pristine** state?

And as I've said earlier:

> After looking at the source code, it looks like markAsPristine() method is very simple and doesn't emit anything.

Please look at the link: https://github.com/angular/angular/blob/master/packages/forms/src/model.ts#L288

I think it would be possible to watch for changes through `ngDoCheck()`, but I can't find a way to actually access the `FormControl` of my component from inside of the component (I would really appreciate any suggestions on this). And even if this works with `ngDoCheck()` are you sure it's the right way to handle this case? Isn't it better to provide an observable for dirty/pristine state or to extend the `ControlValueAccessor` interface? This **is** looking like a valid feature request.

**Comment by Toxicable** [comment]
@slavafomin Sorry, my mistake for some reason I thought you were talking about validity.
>You will need to use subscribe() instead of map(), otherwise cold Observable wouldn't emit anything. At least in my experience.

It was an example hence why I wrote `...`

>I think it would be possible to watch for changes through ngDoCheck()

Please don't do that, you could possibly casue your application to slow down very fast by putting logic in your `ngDoCheck()`
My recommendataion would be using something like this
```
this.control.valueChanges.map( value => this.control.dity)...
```
Which will give you a stream for when the control is dirty, then you can extend it to include the other properties as required.

**Comment by slavafomin** [comment]
@Toxicable I think we all in a hurry these days ; )

I've tried to subscribe to `valueChanges` observable, but it's not emitting anything when markAsPristine() is called either.

And again, could you please give me a hint how to access the `FormControl` associated with my custom form component from inside of it? I'm using FormBuilder and specifying control for the component using `formControlName` attribute.

**Comment by Toxicable** [comment]
to grab a FormControl object you can do this
```
this.form = this.fb.group({
  'name': 'toxicable,
  'age': 94
})
var control = this.form.get('name') //returns AbstractControl which has all the properties you need on it
var control = <FormControl>this.form.get('name') //you can also cast it if you know it's going to be a control
```

**Comment by Toxicable** [comment]
I don't think there is a reasonable way for getting dirty/pristine/etc now that I look at it a bit more. Unless I'm missing something here but I don't believe I am, in which case I think it would be worthy of a Feature request. I'll make a PR for it next week and see what I can come up with.

**Comment by slavafomin** [comment]
Thanks, I think it's very useful feature. Strangely, I'm the first to find out that it's missing.

Considering your example, I'm looking for a way to access `FormControl` from **inside of my custom form component**, not from parent component. Let me explain with the code:

```typescript
@Component({
  selector: 'mbs-date-input',
  templateUrl: './date-input.component.html',
  providers: [{
    provide: NG_VALUE_ACCESSOR,
    useExisting: DateInputComponent,
    multi: true,
  }]
})
export class DateInputComponent implements ControlValueAccessor, OnInit {

  ngOnInit () {
    // How do I access `FormControl` externally associated with my component from here?
  }

}

```

**Comment by Toxicable** [comment]
I have had a few other mentions of it but I have not really looked at it properly till now, also generally people just bind to the values in their view and don't need it pragmatically.

I'm not overly familiar with `ControlValueAccessor` but from what I understand you cant access it. I could be wrong here but from my understanding the FormControl get's your ValueAccessor as an injection, that's what this is setting up here, in which case the FormControl is not available from your accessor.
```
{
    provide: NG_VALUE_ACCESSOR,
    useExisting: DateInputComponent,
    multi: true,
  }
```
So you'd have to wrap FormControl instead :D

**Comment by slavafomin** [comment]
I've studied the code of `FormControlName` directive and it's inspired me with the following solution:

```typescript
@Component({
  selector: 'mbs-date-input',
  templateUrl: './date-input.component.html',
  providers: [{
    provide: NG_VALUE_ACCESSOR,
    useExisting: DateInputComponent,
    multi: true,
  }]
})
export class DateInputComponent implements ControlValueAccessor, OnInit {

  @Input() formControlName: string;

  constructor (
    @Optional() @Host() @SkipSelf()
    private controlContainer: ControlContainer
  ) {
  }


  ngOnInit () {
    const control = this.controlContainer.control.get(this.formControlName);
  }

}
```

I'm injecting the parent `FormGroup` and then getting the control from it using control name obtained through `@Input() formControlName: string;` binding. **Do you think it's a viable approach?**

Also, do you think that inability of accessing the `FormControl` instance directly from the custom form component is a limitation of Angular which needs to be addressed? **Would you like me to file another issue for this?**

**Comment by Toxicable** [comment]
That looks reasonable way to retrieve the FormControl, it dosen't feel right (explained below) but if it's possible why then not? 

>Also, do you think that inability of accessing the FormControl instance directly from the custom form component is a limitation of Angular which needs to be addressed? Would you like me to file another issue for this?

I don't agree with this. Since all your are defining here is how the Control *accesses* it's value not how the control does other things like set validity, statuses, etc...

**Comment by slavafomin** [comment]
So, if it's not a limitation, than the solution must be officially documented, or at least there should be articles or SO answers regarding this, but I've failed to find anything related to this problem. Also, the solution, which I've provided above is also sub-optimal, because it will not work in all cases (e.g. when `formControlName` directive is not used). And in order to make it work in all situations you will need to add a lot of boilerplate code to your component. Shouldn't the framework solve those problems for the developers? I think this is the main reason why we create frameworks in the first place.

So, in my opinion, Angular should consider this as a valid feature request or should regard my use case as invalid (in such case, there must be a documented alternative way).

**Comment by slavafomin** [comment]
I've came up with two possible workarounds for the original issue:

## Monkey-patching the `markAsPristine()`

```typescript
@Component({
  selector: 'my-custom-form-component',
  templateUrl: './custom-form-component.html',
  providers: [{
    provide: NG_VALUE_ACCESSOR,
    useExisting: MyCustomFormComponent,
    multi: true
  }]
})
export class MyCustomFormComponent implements NG_VALUE_ACCESSOR, OnInit {

  private control: AbstractControl;


  ngOnInit () {
    const self = this;
    const origFunc = this.control.markAsPristine;
    this.control.markAsPristine = function () {
      origFunc.apply(this, arguments);
      console.log('Marked as pristine!');
    }
  }

}
```

## Watching for changes with `ngDoCheck`

```typescript
@Component({
  selector: 'my-custom-form-component',
  templateUrl: './custom-form-component.html',
  providers: [{
    provide: NG_VALUE_ACCESSOR,
    useExisting: MyCustomFormComponent,
    multi: true
  }]
})
export class MyCustomFormComponent implements NG_VALUE_ACCESSOR, DoCheck {

  private control: AbstractControl;

  private pristine = true;


  ngDoCheck (): void {
    if (this.pristine !== this.control.pristine) {
      this.pristine = this.control.pristine;
      if (this.pristine) {
        console.log('Marked as pristine!');
      }
    }
  }

}
```

---

I've also updated the questions on StackOveflow:
- [How do I know when custom form control is marked as pristine in Angular?](https://stackoverflow.com/q/44730711/1056679)
- [Get access to FormControl from the custom form component in Angular](https://stackoverflow.com/q/44731894/1056679)

**Comment by kara** [comment]
@Toxicable I'd wait on this one until we make some decisions about refactoring forms to use streams.

**Comment by Toxicable** [comment]
@kara Sounds good to me, let me know how it goes :D

**Comment by BioPhoton** [comment]
Hi @slavafomin 

I have the same problem. And i came up with another approach for listening to pristine dirty or reset actions. I created a Observable that merges status changes and value changes and this one a trransform to the srteam of my interest. In the bleow example im interested in the reset event:
```typescript
Observable.combineLatest(this.control.statusChanges, this.control.valueChanges)
      // called when dir gets destroyed
      .takeUntil(this.destroy$.asObservable())
      .map((n) => {
        return {
          valid: this.control.valid,
          invalid: this.control.invalid,
          touched: this.control.touched,
          untouched: this.control.untouched,
          dirty: this.control.dirty,
          pristine: this.control.pristine,
          status: this.control.status,
          pending: this.control.pending
        }
      })
      .filter((controlState) => {
        const resetState = {
          dirty: false,
          pristine: true,
          touched: false,
          untouched: true
        };

        return Object
          .keys(resetState)
          .reduce((state, item) => {
            return !state ? false : controlState[item] === resetState[item];
          }, true)
      })
      .subscribe((controlState) => {
        this.onResetEvent(controlState);
      })
```

Always interested in feedback!!
Maybe this helps, best
Michael

**Comment by AndrewBogushevich** [comment]
@BioPhoton 
Hi, it perfectly combines streams but still doesn't fired on input blur(touched changed).

**Comment by BioPhoton** [comment]
Thats true @AndrewBogushevich!

I added a `@HostListener('blur', [$event]) `that next my `blur$` subject with `{ touched: true, untouched: false}`
Than i added it to the existing stream:
`Observable.combineLatest(this.control.statusChanges, this.control.valueChanges, this.blur$)`

Should now fire also on onTouched...
Is that a solution for you?

**Comment by BioPhoton** [comment]
To complete it you could do same with `focus$`.

**Comment by BioPhoton** [comment]
Another way would be:
```typescript
registerOnTouched(fn: () => void): void {
    this.onTouched = () => {this.touched$.next(); fn(); };
  }
```

**Comment by gbouteiller** [comment]
@BioPhoton 
Hi, I think it doesn't solve the problem for markAsTouched as there is no observable for that. The only thing that seems to work for now is @slavafomin monkey patching, but as he said, it is not ideal. I think we really should have access to observables for touched/untouched and stuff like that.

**Comment by kara** [comment]
Let's actually collapse this issue with the existing on here (since it also includes touched, etc):

https://github.com/angular/angular/issues/10887

Closing as a dupe - continue any discussion on the other one.

**Comment by dgroh** [comment]
from come the instace of `this.control`?

**Comment by inigay** [comment]
I am in the same situation. Having an event that needs to check validity (which is easy) and then if it's not valid I need to show error messages on my formControl components, which do not trigger when doing `markAsDirty()` that's a pity. Not sure what's the use of these functions then.

**Comment by Olgagr** [comment]
@dgroh I think it's taken from previous example:

```
@Input() formControlName: string;

  constructor (
    @Optional() @Host() @SkipSelf()
    private controlContainer: ControlContainer
  ) {
  }


  ngOnInit () {
    const control = this.controlContainer.control.get(this.formControlName);
  }

```

However, it brings up other problem that you have to mark formControlName as input which is not very handy (I prefer to have it as an attribute).

**Comment by djindjic** [comment]
hey @slavafomin , there was this question above but I need to repeat it as very important, I cannot access `private control: AbstractControl;` (from where is comes?), and also get an compilation error when try to implement `NG_VALUE_ACCESSOR` instead of `ControlValueAccessor`

**Comment by kumaresan-subramani** [comment]
HI @slavafomin ,
 In my case, i have my date-picker component if i change date programmatically inside the form ng-dirty class is getting added but it should not be added that is the behaviour.
In that case, i have to set element to pristine state manually, do you have any idea how to do that....! 

this is my form-base.ts code:
import { EventEmitter, ElementRef, DoCheck } from '@angular/core';
import { getValue, setValue, isNullOrUndefined } from '@syncfusion/ej2-base';
import { ControlValueAccessor, AbstractControl } from '@angular/forms';

/**
 * Angular Form Base Module
 */
export class FormBase<T> implements ControlValueAccessor, DoCheck {
    public value: T;
    public checked: boolean;
    private skipFromEvent: boolean;

    public propagateChange(_: T): void { return; }
    public propagateTouch(): void { return; }
    public enabled: Object;

    public element: HTMLElement;
    public inputElement: HTMLInputElement;
    private ngEle: ElementRef;

    private elem: AbstractControl;
    private pristine = true;
    
    ngDoCheck (): void {
        if (this.pristine !== this.elem.pristine) {
          this.pristine = this.elem.pristine;
          if (this.pristine) {
            console.log('Marked as pristine!');
          }
        }
      }
    
    public appendTo: (ele: string | HTMLElement) => void;

    public focus: EventEmitter<Object>;
    public blur: EventEmitter<Object>;

    public localChange(e: { value?: T, checked?: T }): void {
        let value: T = (e.checked === undefined ? e.value : e.checked);
        if (this.propagateChange !== undefined && value !== undefined) {
            // Update angular from our control
            this.propagateChange(value);
        }
        if ((this.element.classList.contains('ng-untouched')) && (this.element.classList.contains('ng-dirty'))) {
            // this.elem.markAsPristine();
        }
        this.ngDoCheck();
    }

    public properties: Object;
    public saveChanges: Function;


    public registerOnChange(registerFunction: (_: T) => void): void {
        this.propagateChange = registerFunction;
    }

    public registerOnTouched(registerFunction: () => void): void {
        this.propagateTouch = registerFunction;
    }
    public twoWaySetter(newVal: Object, prop: string): void {
        let oldVal: Object = getValue(prop, this.properties);
        let ele: HTMLElement = this.inputElement || this.element;
        if (oldVal === newVal &&
            ((<HTMLInputElement>ele).value === undefined || (<HTMLInputElement>ele).value === '')) {
            return;
        }
        this.saveChanges(prop, newVal, oldVal);
        setValue(prop, (isNullOrUndefined(newVal) ? null : newVal), this.properties);
        getValue(prop + 'Change', this).emit(newVal);
    }
    public ngAfterViewInit(): void {
        // Used setTimeout for template binding
        // Refer Link: https://github.com/angular/angular/issues/6005
        setTimeout(() => {
            /* istanbul ignore else */
            if (typeof window !== 'undefined') {
                this.appendTo(this.element);
                let ele: HTMLElement = this.inputElement || this.element;
                ele.addEventListener('focus', this.ngOnFocus.bind(this));
                ele.addEventListener('blur', this.ngOnBlur.bind(this));
            }
        });
    }
    public setDisabledState(disabled: boolean): void {
        this.enabled = !disabled;
    }

    public writeValue(value: T): void {
        //update control value from angular
        if (this.checked === undefined) {
            this.value = value;
        } else {
            if (typeof value === 'boolean') {
                this.checked = value;
            } else {
                this.checked = value === this.value;
            }
        }
        if (value === null) {
            return;
        }
    }

    public ngOnFocus(e: Event): void {
        /* istanbul ignore else */
        if (this.skipFromEvent !== true) {
            this.focus.emit(e);
        }
    }

    public ngOnBlur(e: Event): void {
        this.propagateTouch();
        /* istanbul ignore else */
        if (this.skipFromEvent !== true) {
            this.blur.emit(e);
        }
    }
}


if i use like this[ your second method ] it sows error like cannot find pristine of undefined
